sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -350,7 +346,9 @@
             Scanner s = new Scanner(connection.getInputStream(), StandardCharsets.UTF_8.name()).useDelimiter("\\A");
             String result = s.hasNext() ? s.next() : "";
 
-            return Mono.just(SERIALIZER_ADAPTER.deserialize(result, MSIToken.class, SerializerEncoding.JSON));
+            MSIToken msiToken = SERIALIZER_ADAPTER.deserialize(result, MSIToken.class, SerializerEncoding.JSON);
+            return Mono.just(new AccessToken(msiToken.getToken(),
+                    msiToken.getExpiresAt().plusMinutes(2).minus(options.getRefreshBeforeExpiry())));
```

Comment:
This adjustment of expires time looks good to me, but I have a question on when should we acquire the token from MSI endpoint. This method seems to retrieve the token as soon as it is called. Let's say it is called at 1 pm so token acquired at 1.00 pm, now let's caller subscribed to the returned Mono at 1.10 pm, now caller is served with 10 min old token. Should we delay the token acquisition until the caller subscribes to the returned mono? (I mean putting the entire acquisition logic in Mono.defer).

Reply:
yes I can move it inside a Mono.defer.